### PR TITLE
[ENH] Allow setting `vmin` in `plot_glass_brain` and `plot_stat_map`

### LINKED ISSUE #1149: Adding back in vmin to plotting colormap command
At the moment the [_get_colorbar_and_data_ranges](https://github.com/nilearn/nilearn/blob/40069cb14b43333a73137797eac895cfb054db29/nilearn/plotting/img_plotting.py#L55) function within [img_plotting.py](https://github.com/nilearn/nilearn/blob/40069cb14b43333a73137797eac895cfb054db29/nilearn/plotting/img_plotting.py) by design forces the stat_map to be shown between `-vmax` and `vmax` or between `0` and `vmax`.

This clearly makes sense for [diverging colormaps](http://matplotlib.org/examples/color/colormaps_reference.html) but it doesn't make sense if you want to show, for example, regions that are significantly positively correlated using the `autumn` map or any other [sequential colormaps](http://matplotlib.org/examples/color/colormaps_reference.html). Ideally I'd like to be able to use the whole range of the colormap for values ranging between `2.3` and `5.0`. As the code currently runs, I'm left with only yellow regions!

encode trial_gt_baseline allwords allsubs groupmean_glassbrain_ortho

What I'd really like to do is to have the colorbar start at `2.3` and finish at `5.0` and for that range to cover the whole colormap.

It looks like it would be easy to change the [_get_colorbar_and_data_ranges](https://github.com/nilearn/nilearn/blob/40069cb14b43333a73137797eac895cfb054db29/nilearn/plotting/img_plotting.py#L55) function to bring back the ability to set lower and upper limits for a colormap? Is there a reason not to?

### LINKED ISSUE #3084: Adding a vmin option to plot_stat_map plot_glass_brain functions
Hi,
I see that there have been a couple of related issues in the past requesting a "vmin" parameter for the plot_stat_map and/or plot_glass_brain functions (e.g. [this](https://github.com/nilearn/nilearn/issues/2165) and [this](https://github.com/nilearn/nilearn/issues/1149)), and was wondering if this has been implemented/is being worked on. 

### Benefits to the change
While the current implementation (only one half of the spectrum for positive/negative values alone) is a great way to ensure that positive/negative activations maintain their separate colors, it will be very helpful to get the whole color spectrum (within a desired vmin/vmax) to present other data (e.g. timepoints at which certain events occur) . This is especially important if the data is skewed (i.e., several nodes have very low/high values). Example below:

image

I'd like to get the whole range of the colormap 'viridis' on the brainmap - i.e., from dark blue to yellow, such as is possible with seaborn heatmaps (example below): 
image

The suggested solutions that I found so far - setting symmetric_cbar to False or using the plot_img function instead of the ones  mentioned above won't suffice unfortunately.

Thanks in advance!

### PR DESCRIPTION
**Pull request by michellewang**
- Closes #3084
- Closes #1149

Changes proposed in this pull request:

- Made `plot_glass_brain` use its `vmin` argument instead of silently ignoring it
  - This involved changing `_get_colorbar_and_data_ranges`, which changes the behaviour of `plot_stat_map` to also allow `vmin` to be set (instead of throwing an error)
- Updated docstring for `symmetric_cbar`, since the colorbar will no longer be restricted to `(-vmax, vmax)`, `(0, vmax)`, `(-vmax, 0)`, and the colormap range will also change depending on `vmin`/`vmax`
- Removed `symmetric_data_range` argument from `_get_colorbar_and_data_ranges`, which was added in #3873 to preserve the old behaviour in these functions. We can still get the symmetric data range behaviour if:
  - `symmetric_cbar` is `True`, or
  - `symmetric_cbar` is `'auto'`, `vmin` or `vmax` is `None`, and the data has both positive and negative values
    - This means that a call with default arguments will plot a symmetric colorbar, same as before
- Refactored test cases for `_get_colorbar_and_data_ranges` to have clearer expected results
- Added some example glass brain plots in `examples/01_plotting/plot_demo_glass_brain_extensive.py`
  - Other glass brain examples still show the same plots

Example `plot_glass_brain` calls with `vmin`:

```python
from nilearn import datasets, plotting

stat_img = datasets.load_sample_motor_activation_image()

plotting.plot_glass_brain(
    stat_img,
    colorbar=True,
    display_mode="lyrz",
    threshold=2,
    cmap='viridis',
    vmin=2,                  # use the whole colormap range even if there is a threshold
    symmetric_cbar=False,
)
```

image

```python
plotting.plot_glass_brain(
    stat_img,
    colorbar=True,
    display_mode="lyrz",
    threshold=2,
    cmap='viridis',
    vmin=-5,
    vmax=8,
    plot_abs=False,
)
```

image

Example `plot_stat_map` calls with `vmin`:

```python
from nilearn import datasets, plotting
stat_img = datasets.load_sample_motor_activation_image()

plotting.plot_stat_map(
    stat_img,
    colorbar=True, 
    threshold=3,
    vmin=-9,   # symmetric from -9 to 9
)
```

image

```python
plotting.plot_stat_map(
    stat_img,
    colorbar=True, 
    threshold=3,
    vmin=0,             # non-negative values only
    vmax=10,
    symmetric_cbar=False,
    cmap='viridis',
)
```

image

```python
plotting.plot_stat_map(
    stat_img,
    colorbar=True, 
    threshold=3,
    vmin=3,            # use the whole colormap range even if there is a threshold
    vmax=10,
    symmetric_cbar=False,
    cmap='viridis',
)
```

image

### CODE REVIEW

**Comment by ymzayek** [comment]
Thx @michellewang for working on this! The example plots look great! Will give a more thorough review soon.

**Review by ymzayek — COMMENTED** [review]
Overall all, looks pretty good! Can you explain the refactoring concerning the plotting of absolute values?

File: nilearn/plotting/displays/_slicers.py [hunk 1]
```diff
@@ -461,15 +451,36 @@ def _map_show(
...
+                    vmax=kwargs.get("vmax"),
+                )
+
                 im = display_ax.draw_2d(
                     data_2d, data_bounds, bounding_box, type=type, **kwargs
                 )
                 ims.append(im)
         return ims
 
+    def _threshold(self, data, threshold=None, vmin=None, vmax=None):
```
[the remark this answers is not in the document]
Comment (michellewang) [reply nilearn/plotting/displays/_slicers.py:470]: Yes it applies the same logic (for the same reason as explained [here](https://github.com/nilearn/nilearn/pull/3873#discussion_r1301785684)), though `surf_plotting._threshold` returns the mask and this returns the masked data array. Do we want to refactor them to use the same function? I wasn't sure how best to do that since the surface plotting functions do not seem to use the Nilearn slicers

**Comment by michellewang** [comment]
> Can you explain the refactoring concerning the plotting of absolute values?

Yes, I had to move the thresholding because it needs to happen after we take the absolute value of the data (which happens in `GlassBrainAxes.transform_to_2d`). Otherwise, when `plot_abs` is `True`, `vmin` is set to `0`, and the new thresholding function will mask all negative values in the data (if a non-zero threshold is defined), so the resulting plot will only show values that were positive in the original data instead of the absolute value of all values.

I'll remove the lines with `data_selection = np.abs(data_selection)` in `nilearn/plotting/displays/_axes.py` because after looking at them again I think they are redundant with what the code already did.

File: nilearn/plotting/displays/_slicers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ymzayek) [reply nilearn/plotting/displays/_slicers.py:471]: > Do we want to refactor them to use the same function?
At this point I'd say no, your reasoning is good.

**Comment by ymzayek** [comment]
> Also: are there any tests for slicers and `_map_show`? I couldn't find any from a quick search

Slicers are tested in `nilearn/plotting/tests/test_displays.py` but `_map_show` is not tested directly (only through calls of `add_contours` and `add_overlay`). I'd suggest adding a test relevant to the changes you made.

File: nilearn/_utils/docs.py [hunk 2]
```diff
@@ -871,16 +871,13 @@ def custom_function(vertices):
...
-
-        The colormap will always range from `-vmax` to `vmax`.
-
+symmetric_cbar : :obj:`bool`, or "auto", default="auto"
+    Specifies whether the colorbar and colormap should range from `-vmax` to
+    `vmax` (or from `vmin` to `-vmin` if `-vmin` is greater than `vmax`) or
+    from `vmin` to `vmax`.
+    Setting to `"auto"` will select the former if either 1) `vmin` or `vmax` is
+    None and the image has both positive and negative values, or 2) `vmin` is
+    equal to `-vmax`.
```
Comment (bthirion) [inline nilearn/_utils/docs.py:880]: If  `vmin`  is   equal to `-vmax`, then the two behaviors are equivalent, no ?

File: nilearn/plotting/img_plotting.py [hunk 3]
```diff
@@ -1073,20 +1077,30 @@ def plot_glass_brain(stat_map_img,
...
     if stat_map_img:
         stat_map_img = _utils.check_niimg_3d(stat_map_img, dtype='auto')
         if plot_abs:
-            cbar_vmin, cbar_vmax, vmin, vmax = _get_colorbar_and_data_ranges(
-                _safe_get_data(stat_map_img, ensure_finite=True),
-                vmax=vmax,
-                symmetric_cbar=symmetric_cbar,
-                force_min_stat_map_value=0)
+            if vmin is not None and vmin < 0:
+                raise ValueError(
```
Comment (bthirion) [inline nilearn/plotting/img_plotting.py:1090]: Wondering if we should really raise en Error here ? A warning could be enough. Think of a loop creating many figures with `plot_abs` set to `True`, but a `vmin` calculated for each figure iteratively.

File: nilearn/_utils/docs.py (hunk 2 above)
Comment (michellewang) [reply nilearn/_utils/docs.py:880]: Yes, you're right. Updated the docstring

File: nilearn/plotting/img_plotting.py (hunk 3 above)
Comment (michellewang) [reply nilearn/plotting/img_plotting.py:1090]: Ok, I can force `vmin` to be `0` if it's negative and raise a warning instead of the error. Does that sound okay?

Comment (bthirion) [reply nilearn/plotting/img_plotting.py]: Why force vmin to 0 ? The user may have his/her reasons to set it to a given value.

Comment (michellewang) [reply nilearn/plotting/img_plotting.py]: Ok I see, that makes sense. Updated to allow arbitrary `vmin` value

**Review by bthirion — COMMENTED** [review]
LGTM. Can you consider codecov complaints ?

File: nilearn/plotting/tests/test_img_plotting/test_plot_glass_brain.py [hunk 4]
```diff
@@ -118,4 +126,12 @@ def test_plot_glass_brain_with_completely_masked_img(display_mode):
 
 def test_plot_glass_brain_vmin_vmax(img_3d_mni):
     """Smoke tests for plot_glass_brain being passed vmin and vmax."""
-    plot_glass_brain(img_3d_mni, vmin=-2, vmax=2)
+    # set plot_abs=False because otherwise negative vmin is not allowed
+    plot_glass_brain(img_3d_mni, vmin=-2, vmax=2, plot_abs=False)
```
Comment (ymzayek) [inline nilearn/plotting/tests/test_img_plotting/test_plot_glass_brain.py:130]: I think with the recent update vmin is allowed so you can revert the change in this test

File: nilearn/plotting/tests/test_img_plotting/test_plot_glass_brain.py [hunk 5]
```diff
@@ -118,4 +126,12 @@ def test_plot_glass_brain_with_completely_masked_img(display_mode):
...
+
+
+def test_plot_glass_brain_negative_vmin_with_plot_abs(img_3d_mni):
+    """Test that error is thrown when plot_abs is True and vmin is negative."""
```
Comment (ymzayek) [inline nilearn/plotting/tests/test_img_plotting/test_plot_glass_brain.py:134]:
```suggestion
    """Test that warning is thrown if plot_abs is True and vmin is negative."""
```

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 6]
```diff
@@ -87,6 +87,36 @@
     stat_img, threshold=3, colorbar=True, plot_abs=True, display_mode="yx"
 )
 
+###############################################################################
```
Comment (ymzayek) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py:90]:
```suggestion
# %%
```
for style consistency

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 7]
```diff
@@ -87,6 +87,36 @@
...
+    colorbar=True,
+    plot_abs=False,
+    display_mode="yz",
+    vmin=0,
+    threshold=2,
+    symmetric_cbar=False,
+    cmap="viridis",
+)
+
+###############################################################################
```
Comment (ymzayek) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py:107]:
```suggestion
# %%
```

**Comment by ymzayek** [comment]
> LGTM. Can you consider codecov complaints ?

Not sure why that is happening. I checked the report and the added code is fully covered and the codecov/patch report is 100% so I think it's ok. Seems like there is a small decrease on the project level but this seems to come from indirect changes. I think @michellewang you can just make the changes from the last review and this should be good to go.

**Comment by michellewang** [comment]
Ok, I was confused by the Codecov report. Just addressed the remaining comments, thanks @ymzayek and @bthirion for reviewing!

**Review by ymzayek — APPROVED** [review]
Just FYI something went wrong with the merge so I redid it. This PR LGTM, thanks @michellewang!

**Review by bthirion — APPROVED** [review]
LGTM, thx.

File: nilearn/plotting/displays/_slicers.py [hunk 8]
```diff
@@ -398,16 +398,6 @@ def _map_show(
...
-        if threshold is not None:
-            data = _safe_get_data(img, ensure_finite=True)
-            if threshold == 0:
-                data = np.ma.masked_equal(data, 0, copy=False)
-            else:
-                data = np.ma.masked_inside(
-                    data, -threshold, threshold, copy=False
-                )
-            img = new_img_like(img, data, img.affine)
-
```
Comment (Remi-Gau) [inline nilearn/plotting/displays/_slicers.py:410]: - removing this causes a regression https://github.com/nilearn/nilearn/issues/4255
